test/lit/passes/type-generalizing.wast

```diff
@@ -274,4 +444,1128 @@
+ ;; CHECK-NEXT:   (local.get $x)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (extern.internalize
+ ;; CHECK-NEXT:   (local.get $var)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $any-convert-extern-non-nullable (param $x (ref noextern)) (result (ref any))
+  (local $var (ref noextern))
+  (local.set $var
+   (local.get $x)
+  )
+  ;; Require that typeof($var) <: (ref extern).
+  (extern.internalize
+   (local.get $var)
+  )
+ )
+
+ ;; CHECK:      (func $extern-convert-any-nullable (type $13) (result externref)
+ ;; CHECK-NEXT:  (local $var anyref)
+ ;; CHECK-NEXT:  (local.set $var
+ ;; CHECK-NEXT:   (ref.i31
+ ;; CHECK-NEXT:    (i32.const 0)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (extern.externalize
+ ;; CHECK-NEXT:   (local.get $var)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $extern-convert-any-nullable (result externref)
+  (local $var (ref i31))
+  (local.set $var
+   (i31.new
+    (i32.const 0)
+   )
+  )
+  ;; Require that typeof($var) <: anyref.
+  (extern.externalize
+   (local.get $var)
+  )
+ )
+
+ ;; CHECK:      (func $extern-convert-any-non-nullable (type $14) (result (ref extern))
+ ;; CHECK-NEXT:  (local $var (ref any))
+ ;; CHECK-NEXT:  (local.set $var
+ ;; CHECK-NEXT:   (ref.i31
+ ;; CHECK-NEXT:    (i32.const 0)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (extern.externalize
+ ;; CHECK-NEXT:   (local.get $var)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $extern-convert-any-non-nullable (result (ref extern))
+  (local $var (ref i31))
+  (local.set $var
+   (i31.new
+    (i32.const 0)
+   )
+  )
+  ;; Require that typeof($var) <: anyref.
+  (extern.externalize
+   (local.get $var)
+  )
+ )
+
+ ;; CHECK:      (func $helper-any_any (type $15) (param $0 anyref) (param $1 anyref)
+ ;; CHECK-NEXT:  (unreachable)
+ ;; CHECK-NEXT: )
+ (func $helper-any_any (param anyref anyref)
+  (unreachable)
+ )
+)
+
+(module
+ ;; CHECK:      (type $top (sub (func (param i31ref) (result anyref))))
+ (type $top (sub (func (param i31ref) (result anyref))))
+ ;; CHECK:      (type $mid (sub $top (func (param eqref) (result anyref))))
+ (type $mid (sub $top (func (param eqref) (result anyref))))
+ ;; CHECK:      (type $2 (func (result eqref)))
+
+ ;; CHECK:      (type $bot (sub $mid (func (param eqref) (result eqref))))
+ (type $bot (sub $mid (func (param eqref) (result eqref))))
+
+ ;; CHECK:      (type $4 (func (result anyref)))
+
+ ;; CHECK:      (func $call-ref-params-limited (type $4) (result anyref)
+ ;; CHECK-NEXT:  (local $f (ref null $mid))
+ ;; CHECK-NEXT:  (local $arg eqref)
+ ;; CHECK-NEXT:  (call_ref $mid
+ ;; CHECK-NEXT:   (local.get $arg)
+ ;; CHECK-NEXT:   (local.get $f)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $call-ref-params-limited (result anyref)
+  (local $f (ref null $bot))
+  (local $arg i31ref)
+  ;; Require that typeof($f) <: $mid and that typeof($arg) <: eqref. In
+  ;; principle we could generalize $f up to $top, but then we wouldn't be able
+  ;; to generalize $arg at all.
```

Comment:
Why is it better to partially generalize both rather than fully generalize `$f` and leave `$arg` as is?

Reply:
We could construct situations in which generalizing the function reference is more beneficial and we could construct situations in which generalizing the parameters is more beneficial. It's unclear which will be more important in practice, so we should definitely experiment with it. For now, splitting the difference seemed like a good conservative starting point.